### PR TITLE
[BEAM-8550] Requires time sorted input

### PR DESCRIPTION
**Pull request by je-ik**
Proposal of [BEAM-8550].

[1] https://docs.google.com/document/d/1ObLVUFsf1NcG8ZuIZE4aVy2RYKx2FfyMhkZYWPnI9-c/edit?usp=sharing

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | --- | ---
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_Verify/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python3_Verify/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/) <br> [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/)

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [Build Status](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) 
Portable | --- | [Build Status](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

### CODE REVIEW

**Comment by je-ik** [comment]
Run Direct ValidatesRunner

**Comment by je-ik** [comment]
Run Dataflow ValidatesRunner

**Comment by je-ik** [comment]
Run Flink ValidatesRunner

**Comment by je-ik** [comment]
Run Spark ValidatesRunner

**Comment by je-ik** [comment]
Run Java Flink PortableValidatesRunner Streaming

**Comment by je-ik** [comment]
Run Java Flink PortableValidatesRunner Batch

**Comment by je-ik** [comment]
Run Java PreCommit

**Comment by je-ik** [comment]
Run Python2_PVR_Flink PreCommit

**Comment by je-ik** [comment]
Run Java Flink PortableValidatesRunner Streaming

**Comment by je-ik** [comment]
Run Direct ValidatesRunner

**Comment by je-ik** [comment]
Run Dataflow ValidatesRunner

**Comment by je-ik** [comment]
Run Samza ValidatesRunner

**Comment by je-ik** [comment]
Run Java Flink PortableValidatesRunner Streaming

**Comment by je-ik** [comment]
Run Java Flink PortableValidatesRunner Batch

**Comment by je-ik** [comment]
Run Flink ValidatesRunner

**Comment by je-ik** [comment]
Run Spark ValidatesRunner

**Comment by je-ik** [comment]
Run Portable_Python PreCommit

**Comment by je-ik** [comment]
Run Flink ValidatesRunner

**Comment by je-ik** [comment]
Run Java Flink PortableValidatesRunner Streaming

**Comment by je-ik** [comment]
Run Java Flink PortableValidatesRunner Batch

**Comment by je-ik** [comment]
Run Flink ValidatesRunner

**Comment by je-ik** [comment]
Run Java Flink PortableValidatesRunner Streaming

**Comment by je-ik** [comment]
Run Java_Examples_Dataflow PreCommit

**Comment by je-ik** [comment]
Run Java PreCommi

**Comment by je-ik** [comment]
Run Java PreCommit

**Comment by je-ik** [comment]
Run Flink ValidatesRunner

**Comment by je-ik** [comment]
Run Direct ValidatesRunner

**Comment by je-ik** [comment]
Run Java Flink PortableValidatesRunner Streaming

**Comment by je-ik** [comment]
Run Java Flink PortableValidatesRunner Batch

**Comment by je-ik** [comment]
Run Flink ValidatesRunner

**Comment by je-ik** [comment]
Run Direct ValidatesRunner

**Comment by je-ik** [comment]
Run Java Flink PortableValidatesRunner Streaming

**Comment by je-ik** [comment]
Run Java Flink PortableValidatesRunner Batch

**Comment by je-ik** [comment]
Run Direct ValidatesRunner

File: runners/core-java/src/main/java/org/apache/beam/runners/core/StatefulDoFnRunner.java [hunk 1]
```diff
@@ -85,45 +123,83 @@ public void startBundle() {
...
+      sortBuffer.add(value);
+      Instant minStamp =
+          MoreObjects.firstNonNull(minStampState.read(), BoundedWindow.TIMESTAMP_MAX_VALUE);
+      if (value.getTimestamp().isBefore(minStamp)) {
+        minStamp = value.getTimestamp();
+        minStampState.write(minStamp);
+        setupFlushTimerAndWatermarkHold(namespace, minStamp);
       }
+    } else {
+      reportDroppedElement(value, window);
```
Comment (JozoVilcek) [inline runners/core-java/src/main/java/org/apache/beam/runners/core/StatefulDoFnRunner.java]: Such internal detail needs to be more explicitly mentioned in javadoc for the annotation.

What if I am processing stream and my out of order spread can be quite large. Does it make sense for me to use this annotation or better not. Some recommendations / hints for users would be great.

This makes me hesitant to use this annotation for stream processing (not that I have use case for it right now) unless I can be sure about timing of my data stream and watermark move.
I wonder if it would make sense to give some tradeoffs between latency, buffer size, lateness and introduce a possibility to hold watermark back for some delta, smaller than max allowed lateness.

File: runners/core-java/src/main/java/org/apache/beam/runners/core/StatefulDoFnRunner.java [hunk 2]
```diff
@@ -85,45 +123,83 @@ public void startBundle() {
...
-    Instant inputWM = cleanupTimer.currentInputWatermarkTime();
+    Instant inputWM = stepContext.timerInternals().currentInputWatermarkTime();
     return gcTime.isBefore(inputWM);
   }
 
+  private void reportDroppedElement(WindowedValue<InputT> value, BoundedWindow window) {
+    droppedDueToLateness.inc();
+    WindowTracing.debug(
+        "StatefulDoFnRunner.processElement: Dropping element at {}; window:{} "
+            + "since too far behind inputWatermark:{}",
```
Comment (JozoVilcek) [inline runners/core-java/src/main/java/org/apache/beam/runners/core/StatefulDoFnRunner.java]: Is this message correct in case of processing and dropping from ordered code path?

File: runners/core-java/src/main/java/org/apache/beam/runners/core/StatefulDoFnRunner.java [hunk 3]
```diff
@@ -132,16 +208,56 @@ public void onTimer(
...
       } else {
         doFnRunner.onTimer(timerId, window, timestamp, timeDomain);
       }
     }
   }
 
-  @Override
-  public void finishBundle() {
-    doFnRunner.finishBundle();
+  // this needs to be optimized (Sorted Map State)
```
Comment (JozoVilcek) [inline runners/core-java/src/main/java/org/apache/beam/runners/core/StatefulDoFnRunner.java]: Is it a performace implication worth to mention to annotation users? 

Can it be link it to a JIRA?

File: runners/direct-java/src/main/java/org/apache/beam/runners/direct/QuiescenceDriver.java [hunk 4]
```diff
@@ -164,7 +164,7 @@ private void fireTimers() {
                     transformTimers.getKey(),
                     (PCollection)
                         Iterables.getOnlyElement(
-                            transformTimers.getExecutable().getInputs().values()))
+                            transformTimers.getExecutable().getMainInputs().values()))
```
[the remark this answers is not in the document]
Comment (je-ik) [reply runners/direct-java/src/main/java/org/apache/beam/runners/direct/QuiescenceDriver.java]: This actually fixes a bug. The original code fails when there are side inputs, because `getInputs` throws exception.

File: runners/direct-java/src/test/java/org/apache/beam/runners/direct/StatefulParDoEvaluatorFactoryTest.java [hunk 5]
```diff
@@ -135,25 +139,29 @@ public void windowCleanupScheduled() throws Exception {
...
+              public void process(ProcessContext c) {}
+            },
+            mainOutput,
+            TupleTagList.empty(),
+            Collections.emptyList(),
+            DoFnSchemaInformation.create(),
+            Collections.emptyMap());
+
+    final PCollection<KeyedWorkItem<String, KV<String, Integer>>> grouped;
+    grouped = gbkThenStatefulParDo.groupToKeyedWorkItem(input);
```
[the remark this answers is not in the document]
Comment (je-ik) [reply runners/direct-java/src/test/java/org/apache/beam/runners/direct/StatefulParDoEvaluatorFactoryTest.java]:
```suggestion
    final PCollection<KeyedWorkItem<String, KV<String, Integer>>> grouped =
        gbkThenStatefulParDo.groupToKeyedWorkItem(input);
```

File: runners/direct-java/src/test/java/org/apache/beam/runners/direct/StatefulParDoEvaluatorFactoryTest.java [hunk 6]
```diff
@@ -135,25 +139,29 @@ public void windowCleanupScheduled() throws Exception {
...
-        input
-            .apply(
-                new ParDoMultiOverrideFactory.GbkThenStatefulParDo<>(
-                    new DoFn<KV<String, Integer>, Integer>() {
-                      @StateId(stateId)
-                      private final StateSpec<ValueState<String>> spec =
-                          StateSpecs.value(StringUtf8Coder.of());
+    final ParDoMultiOverrideFactory.GbkThenStatefulParDo<String, Integer, Integer>
+        gbkThenStatefulParDo;
+    gbkThenStatefulParDo =
```
[the remark this answers is not in the document]
Comment (je-ik) [reply runners/direct-java/src/test/java/org/apache/beam/runners/direct/StatefulParDoEvaluatorFactoryTest.java]: That actually formats really ugly, so I prefer this one.

File: runners/spark/src/main/java/org/apache/beam/runners/spark/translation/TransformTranslator.java [hunk 7]
```diff
@@ -448,18 +458,152 @@ public String toNativeString() {
...
+  private static <K, V, OutputT>
+      PairFlatMapFunction<Iterator<Tuple2<ByteArray, byte[]>>, TupleTag<?>, WindowedValue<?>>
+          wrapDoFnFromSortedRDD(
+              MultiDoFnFunction<KV<K, V>, OutputT> doFnFunction,
+              Coder<K> keyCoder,
+              Coder<WindowedValue<V>> wvCoder) {
+
+    return (Iterator<Tuple2<ByteArray, byte[]>> in) -> {
+      Iterator<Iterator<Tuple2<TupleTag<?>, WindowedValue<?>>>> mappedGroups;
+      mappedGroups =
```
[the remark this answers is not in the document]
Comment (je-ik) [reply runners/spark/src/main/java/org/apache/beam/runners/spark/translation/TransformTranslator.java]: The reason is formatting. The latter implies additional level of indentation.

File: sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/ParDoTest.java [hunk 8]
```diff
@@ -2363,6 +2367,136 @@ public int hashCode() {
...
+        }
+      }
+      testTimeSortedInput(
+          numElements - 100,
+          numElements - 1,
+          pipeline.apply(input.advanceWatermarkToInfinity()),
+          // Only direct runner is guaranteed to correctly drop elements after watermark  coming
+          // from different (sub-)partitions of TestStream. Cannot reference DirectRunner by
+          // class to avoid cyclic dependencies.
+          pipeline.getOptions().getRunner().getSimpleName().equals("DirectRunner"));
```
[the remark this answers is not in the document]
Comment (je-ik) [reply sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/ParDoTest.java]: Other runners (flink to be specific) might handle watermark differently. Only direct runner synchronously advances watermark, so that what you mark as "late" in TestStream will really end up late. When running the same TestStream with flink, the watermark might get updated at different times and thus the data might not be late and the output is indeed different.

File: runners/core-java/src/main/java/org/apache/beam/runners/core/StatefulDoFnRunner.java [hunk 9]
```diff
@@ -85,45 +123,83 @@ public void startBundle() {
...
-        droppedDueToLateness.inc();
-        WindowTracing.debug(
-            "StatefulDoFnRunner.processElement: Dropping element at {}; window:{} "
-                + "since too far behind inputWatermark:{}",
-            input.getTimestamp(),
-            window,
-            cleanupTimer.currentInputWatermarkTime());
+        reportDroppedElement(value, window);
+      } else if (requiresTimeSortedInput) {
+        processElementOrdered(window, value);
```
Comment (je-ik) [inline runners/core-java/src/main/java/org/apache/beam/runners/core/StatefulDoFnRunner.java]: This implies that we have to buffer elements up to watermark + allowed lateness.

File: runners/core-java/src/main/java/org/apache/beam/runners/core/StatefulDoFnRunner.java (hunk 2 above)
Comment (je-ik) [reply runners/core-java/src/main/java/org/apache/beam/runners/core/StatefulDoFnRunner.java]: Should be. But we need to modify the flushing behavior to respect allowed lateness. This is wrong in the current implementation. Will fix that.

File: runners/core-java/src/main/java/org/apache/beam/runners/core/StatefulDoFnRunner.java (hunk 3 above)
Comment (je-ik) [reply runners/core-java/src/main/java/org/apache/beam/runners/core/StatefulDoFnRunner.java]: Actually, it turns out that when the buffer is flushed not per single element, but for each input watermark move, this optimization might be negligible. I'm seeing good performance even without it. JIRA would be good, as when there is Sorted Map State, then this can be leveraged.

Comment (je-ik) [reply runners/core-java/src/main/java/org/apache/beam/runners/core/StatefulDoFnRunner.java]: I will create the JIRA after this is merged.

File: runners/direct-java/src/main/java/org/apache/beam/runners/direct/QuiescenceDriver.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (JozoVilcek) [reply runners/direct-java/src/main/java/org/apache/beam/runners/direct/QuiescenceDriver.java]: To get a better understanding, this is a big triggered changes of this feature. It does not manifest without them, that is why it does not make sense to track it in separate JIRA?

File: runners/spark/src/main/java/org/apache/beam/runners/spark/translation/TransformTranslator.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (JozoVilcek) [reply runners/spark/src/main/java/org/apache/beam/runners/spark/translation/TransformTranslator.java]: imho it is quite bad to have to make such tradeoffs when formatting is working against us. But OK, does not have to concert this PR

File: runners/direct-java/src/main/java/org/apache/beam/runners/direct/QuiescenceDriver.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (je-ik) [reply runners/direct-java/src/main/java/org/apache/beam/runners/direct/QuiescenceDriver.java]: I think it manifests, I just didn't have time to file a JIRA for that. The change is pretty much small, just differentiates "main" input from "any other inputs". `getInputs` returns all inputs including side inputs (as it is today), while `getMainInputs()` return only the non-side inputs, which should be single input in this case and so `Iterables.getOnlyElement` will not fail.

File: sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/ParDoTest.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (JozoVilcek) [reply sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/ParDoTest.java]: Is this a problem of runner (Flink) implementing TestStream sub-optimally or that Beam can not have a full control over watermark behavior? I think it is latter.
From comment, I was not sure if it is workaround of not testing on all runners because it is broken or a workaround of assigning which runner this make sense to run on.

[the remark this answers is not in the document]
Comment (je-ik) [reply sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/ParDoTest.java]: Yes, it is the latter. Direct runner is somewhat special, because it is intentionally designed to be single-process and thus can have stricter guarantees. Another option would be to split (basically copy & paste) this test to two versions - one that would reside in tests specific for DirectRunner and the other would be validates runner test. I think that this is probably not worth the code duplication.

[the remark this answers is not in the document]
Comment (JozoVilcek) [reply sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/ParDoTest.java]: If it is about code duplication around setting up TestStream, that can be reused via shared routine. I would actually prefer cleaner association of which test logic make sense for which test category and avoid "sniffing context" where it is running. It did confuse at least myself.

[the remark this answers is not in the document]
Comment (je-ik) [reply sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/ParDoTest.java]: I'm afraid, that code reuse between the two test modules would not be straightforward and would introduce much more counter intuitive code.
The only option would be to split the test into two parts:
 a) validate that the logic of the late data dropping is correct (that is, test the annotation itself) (this would have to be placed outside validates runner suites)
 b) validate that runners obey the annotation (validates runner)
There would be some code duplication probably, but it might be worth it. I will look into that.

File: runners/core-java/src/main/java/org/apache/beam/runners/core/StatefulDoFnRunner.java (hunk 1 above)
Comment (je-ik) [reply runners/core-java/src/main/java/org/apache/beam/runners/core/StatefulDoFnRunner.java]: I think that this implies that we have to take better care of allowed lateness here. Elements that are more late than allowed lateness are dropped by any stateful dofn, this is no new behavior.

File: runners/direct-java/src/test/java/org/apache/beam/runners/direct/StatefulParDoEvaluatorFactoryTest.java [hunk 10]
```diff
@@ -188,15 +196,35 @@ public void process(ProcessContext c) {}
     // A single bundle with some elements in the global window; it should register cleanup for the
     // global window state merely by having the evaluator created. The cleanup logic does not
     // depend on the window.
-    CommittedBundle<KV<String, Integer>> inputBundle =
+    CommittedBundle<KeyedWorkItem<String, KV<String, Integer>>> inputBundle =
```
[the remark this answers is not in the document]
Comment (je-ik) [reply runners/direct-java/src/test/java/org/apache/beam/runners/direct/StatefulParDoEvaluatorFactoryTest.java]: This relates to how coders are passed around in DirectRunner. See commit dfe825284b8be76736155e82c8ce278a45640595.

File: runners/spark/src/main/java/org/apache/beam/runners/spark/translation/TransformTranslator.java [hunk 11]
```diff
@@ -448,18 +458,152 @@ public String toNativeString() {
...
+      @Override
+      public int numPartitions() {
+        return partitioner.numPartitions();
+      }
+
+      @Override
+      public int getPartition(Object o) {
+        ByteArray b = (ByteArray) o;
+        return partitioner.getPartition(
+            new ByteArray(Arrays.copyOfRange(b.getValue(), 0, b.getValue().length - 8)));
```
[the remark this answers is not in the document]
Comment (je-ik) [reply runners/spark/src/main/java/org/apache/beam/runners/spark/translation/TransformTranslator.java]: I wouldn't do that. The utility class is named `CoderHelpers` and this creates `Partitioner`, which has nothing to do with `Coders`. We would have to refactor it to something like `ByteArrayHelpers`, but I don't think that it would do this PR more readable. We might postpone that after time this is merged.

File: runners/spark/src/main/java/org/apache/beam/runners/spark/translation/TransformTranslator.java [hunk 12]
```diff
@@ -448,18 +458,152 @@ public String toNativeString() {
...
+    return new AbstractIterator<Iterator<WindowedValue<KV<K, V>>>>() {
+
+      @Nullable Tuple2<ByteArray, byte[]> read = null;
+
+      @Override
+      protected Iterator<WindowedValue<KV<K, V>>> computeNext() {
+        readNext();
+        if (read != null) {
+          byte[] value = read._1().getValue();
+          byte[] keyPart = Arrays.copyOfRange(value, 0, value.length - 8);
```
[the remark this answers is not in the document]
Comment (je-ik) [reply runners/spark/src/main/java/org/apache/beam/runners/spark/translation/TransformTranslator.java]: Same problem. Not related to `Coders`.

File: runners/spark/src/main/java/org/apache/beam/runners/spark/translation/TransformTranslator.java [hunk 13]
```diff
@@ -448,18 +458,152 @@ public String toNativeString() {
...
+
+      private Iterator<WindowedValue<KV<K, V>>> createIteratorForKey(byte[] keyPart, K key) {
+
+        return new AbstractIterator<WindowedValue<KV<K, V>>>() {
+          @Override
+          protected WindowedValue<KV<K, V>> computeNext() {
+            readNext();
+            if (read != null) {
+              byte[] value = read._1().getValue();
+              byte[] prefix = Arrays.copyOfRange(value, 0, value.length - 8);
```
[the remark this answers is not in the document]
Comment (je-ik) [reply runners/spark/src/main/java/org/apache/beam/runners/spark/translation/TransformTranslator.java]: I don't folow. Can you elaborate?

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/DoFn.java [hunk 14]
```diff
@@ -656,6 +656,24 @@ public Duration getAllowedTimestampSkew() {
...
+   * alternative user supplied ordering function can be supplied.
+   *
+   * <p>Note that this annotation makes sense only for stateful {@code ParDo}s, because outcome of
+   * stateless functions cannot depend on the ordering.
+   */
+  @Documented
+  @Experimental
+  @Retention(RetentionPolicy.RUNTIME)
+  @Target(ElementType.METHOD)
+  public @interface RequiresTimeSortedInput {}
```
[the remark this answers is not in the document]
Comment (je-ik) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/DoFn.java]: There should be no change regarding handling late data between using and not using this annotation. I will create tests to make sure it is the case.

File: sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/ParDoTest.java [hunk 15]
```diff
@@ -2363,6 +2367,136 @@ public int hashCode() {
...
+      TestStream.Builder<Long> input = TestStream.create(VarLongCoder.of());
+      for (Long stamp : eventStamps) {
+        input = input.addElements(TimestampedValue.of(stamp, Instant.ofEpochMilli(stamp)));
+        if (stamp == 100) {
+          // advance watermark when we have 100 remaining elements
+          // all the rest are going to be late elements
+          input = input.advanceWatermarkTo(Instant.ofEpochMilli(stamp));
+        }
+      }
+      testTimeSortedInput(
```
[the remark this answers is not in the document]
Comment (je-ik) [reply sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/ParDoTest.java]: That is not possible. Every runner might drop different elements, because it depends on how watermark is propagated from the `TestStream` to the downstream operators.

**Comment by je-ik** [comment]
@JozoVilcek I have resolved some issues mentioned. I have come across issue in DirectRunner (not dropping late data), which when resolved might enable adding more tests verifying that the annotation does not imply dropping more data.

File: runners/core-java/src/main/java/org/apache/beam/runners/core/StatefulDoFnRunner.java (hunk 9 above)
Comment (je-ik) [reply runners/core-java/src/main/java/org/apache/beam/runners/core/StatefulDoFnRunner.java]: Fixed.

File: runners/core-java/src/main/java/org/apache/beam/runners/core/StatefulDoFnRunner.java (hunk 1 above)
Comment (je-ik) [reply runners/core-java/src/main/java/org/apache/beam/runners/core/StatefulDoFnRunner.java]: There will be no difference in dropped data when using and not using this annotation. Mailing list thread for reference: https://lists.apache.org/thread.html/c37dfb6c545fba7d794a13c507dccebb654bbd8b317dab748a6775dc%40%3Cdev.beam.apache.org%3E

File: runners/core-java/src/main/java/org/apache/beam/runners/core/StatefulDoFnRunner.java (hunk 2 above)
Comment (je-ik) [reply runners/core-java/src/main/java/org/apache/beam/runners/core/StatefulDoFnRunner.java]: Fixed.

**Comment by je-ik** [comment]
Run Flink ValidatesRunner

File: runners/core-java/src/main/java/org/apache/beam/runners/core/StatefulDoFnRunner.java (hunk 1 above)
Comment (je-ik) [reply runners/core-java/src/main/java/org/apache/beam/runners/core/StatefulDoFnRunner.java]: The conversation there confirms that there is on purpose different behavior of late data dropping than I expected. I will describe this in the javadoc.

Comment (je-ik) [reply runners/core-java/src/main/java/org/apache/beam/runners/core/StatefulDoFnRunner.java]: Added mention to javadoc regarding the change in late data dropping behavior.

**Comment by je-ik** [comment]
Retest this please

**Comment by JozoVilcek** [comment]
lgtm

**Comment by je-ik** [comment]
retest this please

**Comment by je-ik** [comment]
Run Java_Examples_Dataflow PreCommit

**Comment by je-ik** [comment]
Run Direct ValidatesRunner

**Comment by je-ik** [comment]
Run Spark ValidatesRunner

**Comment by je-ik** [comment]
Run Flink ValidatesRunner

**Comment by je-ik** [comment]
Run Java Flink PortableValidatesRunner Streaming

**Comment by je-ik** [comment]
Run Java Flink PortableValidatesRunner Batch

**Comment by je-ik** [comment]
Run Java_Examples_Dataflow PreCommit

**Comment by je-ik** [comment]
Run Java_Examples_Dataflow PreCommit

File: runners/core-java/src/main/java/org/apache/beam/runners/core/StatefulDoFnRunner.java (hunk 3 above)
Comment (je-ik) [reply runners/core-java/src/main/java/org/apache/beam/runners/core/StatefulDoFnRunner.java]: Created https://issues.apache.org/jira/browse/BEAM-9256

File: runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/ExecutableStageDoFnOperator.java [hunk 16]
```diff
@@ -938,4 +938,24 @@ private static void initializeUserState(
     @ProcessElement
     public void doNothing(ProcessContext context) {}
   }
+
+  private static class Locker implements AutoCloseable {
+
+    public static Locker locked(Lock lock) {
+      Locker locker = new Locker(lock);
```
Comment (mxm) [inline runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/ExecutableStageDoFnOperator.java]: What has this change to do with this PR? This creates a new locking object for every state request. Why is this necessary? The existing pattern was sufficient and fast:

```java
try {
lock()
} finally {
unlock()
}
```

The change adds verbosity and touches just touches a lot of code without any reason other than a style change. If at all it adds a regression.

Such unrelated changes should not have been part of this PR.

CC @tweise

Comment (mxm) [reply runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/ExecutableStageDoFnOperator.java]: Besides there are a bunch of other changes which are not related. Please factor those out in individual commits or even into separate PRs.

Comment (je-ik) [reply runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/ExecutableStageDoFnOperator.java]: I agree, that these changes are not directly related to this PR and we can revert them, if necessary.
If I recall correctly, these changes were part of debugging of state related to [FLINK-12653](https://jira.apache.org/jira/browse/FLINK-12653). Just to add context, why these changes are there.
Regarding performance, I think that this change cannot have any measurable impact on performance, because creating object in eden space is pretty cheap in current JVMs.
I agree that the `try` - `finally` is working just fine, the locker pattern is just a safeguard to prevent someone forgetting add the `finally` clause.
But as I said, if these changes are undesirable, we can revert them.
